test/test_scatter_gather_ops.py

```diff
@@ -264,6 +264,46 @@ def test_scatter_reduce_amin(self, device, dtype):
                     expected_result[2] = 0
                 self.assertEqual(input, expected_result)
 
+    @onlyCPU
+    @dtypes(torch.float32, torch.float64, torch.bfloat16)
+    def test_scatter_expanded_index(self, device, dtype):
+        def helper(input_size, idx_size):
+            input = torch.randn(input_size, device=device).to(dtype=dtype)
+            input2 = input.clone()
+
+            shape = [1] * len(input_size)
+            shape[0] = idx_size
+            dim_size = input_size[0]
+            idx = torch.randint(0, dim_size, shape)
+
+            # The fast path on scatter when index is expanded
+            # will depend on sorted index where the collected src indice
+            # for each row in input will be mapped to rowptrs in a CSR format.
+            # Create some empty rows by masking:
+            mask = (idx > 1) * (idx < 4)
+            idx[mask] = 0
+
+            expanded_shape = input_size
+            expanded_shape[0] = idx_size
+            idx = idx.expand(expanded_shape)
+            idx2 = idx.contiguous()
+            src = torch.randn(expanded_shape, device=device).to(dtype=dtype)
+
+            out = input.scatter_add(0, idx, src)
+            out2 = input2.scatter_add(0, idx2, src)
+
+            self.assertEqual(out, out2)
+
+            for reduce in ["sum", "prod", "mean", "amax", "amin"]:
+                for include_self in [True, False]:
+                    out = input.scatter_reduce(0, idx, src, reduce=reduce, include_self=include_self)
+                    out2 = input2.scatter_reduce(0, idx2, src, reduce=reduce, include_self=include_self)
+                    self.assertEqual(out, out2)
```

Comment:
Is there a reason this can't be an OpInfo test?

Reply:
it can be done in OpInfo test but I would rather leave it as a standalone test case in `test_scatter_ops.py`. This is because the usage scenario for `scatter_add` or `scatter_reduce` is actually very **special** where the index tensor is expaned. It is not a common usage scenario but it is very heavily used in PyG for message passing.

Also I need to explicitly create the cases when there are `duplicated rows` in the index tensor ( this correponds to the situation that `each src node has multiple dst nodes`). This is the major difficulty of this PR: I want to parallel on the outer dimension so as to get good performance but directly paralleling on the outer dimension would lead to write conflicts ( since different dst nodes may want to update the same src node from different threads). This is why I introduce the sorting, from the algorithm level, this is equivalent to: firstly convert COO format to CSR format and then do rowwise reduction on a CSR format (CSR has fast rowwise access).

Reply:
ok, maybe worth adding what you wrote here as a comment here